### PR TITLE
feat(headless): Add IP Tab controller

### PR DESCRIPTION
**Pull request by aelfaleh**
- https://coveord.atlassian.net/browse/SVCC-2209

### CODE REVIEW

**Comment by nathanlb** [comment]
Looks like there are more changes here than just the tab controller. I think it would be helpful to reviewers if you could add some more info in the PR description.

File: packages/headless/src/controllers/insight/query-summary/headless-insight-query-summary.ts [hunk 1]
```diff
@@ -1,32 +1,17 @@
...
-}
-
-/**
- * A scoped and simplified part of the headless state that is relevant to the `InsightQuerySummary` controller.
- */
-export interface InsightQuerySummaryState extends CoreQuerySummaryState {}
-
+export type {QuerySummary, QuerySummaryState};
 /**
  * Creates a `InsightQuerySummary` controller instance.
```
Comment (nathanlb) [inline packages/headless/src/controllers/insight/query-summary/headless-insight-query-summary.ts:10]: Like with the others we should change the name in the jsdoc as well to be consistent.

**Review by nathanlb — APPROVED** [review]

File: packages/headless/src/controllers/core/tab/headless-core-tab.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  id: requiredNonEmptyString,
+});
+
+const initialStateSchema = new Schema({
+  isActive: new BooleanValue(),
+});
+
+export interface TabProps {
+  /**
+   * The options for the `CoreTab` controller.
```
Comment (lbergeron) [inline packages/headless/src/controllers/core/tab/headless-core-tab.ts:57]:
```suggestion
   * The options for the `Tab` controller.
```

File: packages/headless/src/controllers/core/tab/headless-core-tab.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * The `Tab` headless controller offers a high-level interface for designing a common tab UI controller.
+ */
+export interface Tab extends Controller {
+  /**
+   * Activates the tab.
+   */
+  select(): void;
+
+  /**
+   * The state of the `CoreTab` controller.
```
Comment (lbergeron) [inline packages/headless/src/controllers/core/tab/headless-core-tab.ts:76]:
```suggestion
   * The state of the `Tab` controller.
```

File: packages/headless/src/controllers/insight/tab/headless-insight-tab.test.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      tab.select();
+
+      expect(engine.actions).toContainEqual(updateActiveTab(id));
+    });
+
+    it('dispatches #executeSearch', () => {
+      tab.select();
+      const action = engine.findAsyncAction(executeSearch.pending);
+      expect(action).toBeTruthy();
+    });
```
Comment (lbergeron) [inline packages/headless/src/controllers/insight/tab/headless-insight-tab.test.ts:85]: I suggest we keep only this test. The others are all duplicates of the core test suite.

**Review by lbergeron — APPROVED** [review]
It looks quite good!

File: packages/headless/src/controllers/insight/tab/headless-insight-tab.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  TabInitialState,
+  TabOptions,
+} from '../../core/tab/headless-core-tab';
+import {InsightEngine} from '../../../app/insight-engine/insight-engine';
+import {logInsightInterfaceChange} from '../../../features/analytics/analytics-actions';
+import {executeSearch} from '../../../features/insight-search/insight-search-actions';
+
+export type {Tab, TabProps, TabState, TabInitialState, TabOptions};
+/**
+ * Creates an insight `Tab` controller instance.
```
[the remark this answers is not in the document]
Comment (nathanlb) [reply packages/headless/src/controllers/insight/tab/headless-insight-tab.ts:15]: If you want to add it in your other renaming PR then I'm on board :)

**Review by btaillon — APPROVED** [review]

File: packages/headless/src/controllers/core/tab/headless-core-tab.test.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      },
+      initialState: {
+        isActive: false,
+      },
+    };
+
+    initTab();
+  });
+
+  it('it adds the correct reducers to engine', () => {
```
Comment (erocheleau) [inline packages/headless/src/controllers/core/tab/headless-core-tab.test.ts:35]:
```suggestion
  it('adds the correct reducers to engine', () => {
```

File: packages/headless/src/controllers/core/tab/headless-core-tab.test.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    it('when isActive is true, it dispatches #updateActiveTab', () => {
+      const {id} = props.options;
+      props.initialState!.isActive = true;
+      initTab();
+
+      expect(engine.actions).toContainEqual(updateActiveTab(id));
+    });
+
+    it('does not throw if initialState is undefined', () => {
```
Comment (erocheleau) [inline packages/headless/src/controllers/core/tab/headless-core-tab.test.ts:69]: should we test that it initializes correctly and not just that it doesn't throw?

File: packages/headless/src/controllers/core/tab/headless-core-tab.test.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      expect(() => initTab()).toThrow('Check the options of buildTab');
+    });
+  });
+
+  describe('#select', () => {
+    it('dispatches #updateActiveTab', () => {
+      const {id} = props.options;
+      tab.select();
+
+      expect(engine.actions).toContainEqual(updateActiveTab(id));
```
Comment (erocheleau) [inline packages/headless/src/controllers/core/tab/headless-core-tab.test.ts:96]: should you expect that the state isActive is also true?

File: packages/headless/src/controllers/core/tab/headless-core-tab.test.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    });
+
+    it('when options #expression is an invalid value, it throws an error', () => {
+      props.options.expression = 1 as unknown as string;
+      expect(() => initTab()).toThrow('Check the options of buildTab');
+    });
+  });
+
+  describe('#select', () => {
+    it('dispatches #updateActiveTab', () => {
```
Comment (erocheleau) [inline packages/headless/src/controllers/core/tab/headless-core-tab.test.ts:92]: might be a nice test also to validate that when another tab was already active, after select() then this tab isActive and the other tab is no longer active?

**Review by erocheleau — CHANGES_REQUESTED** [review]
Just small changes then good to go!

**Review by erocheleau — APPROVED** [review]

**Pull request merged**